File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.py
from collections import deque

def bfs(graph, queue, tar_x, tar_y, visited):
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    
    while queue:
        cur_x, cur_y = queue.popleft()
        if cur_x == tar_x and cur_y == tar_y:
            return visited[cur_x][cur_y]  // 2
        for i in range(4):
            new_x = cur_x + dx[i]
            new_y = cur_y + dy[i]
            if graph[new_x][new_y] == 1 and visited[new_x][new_y] == 1:
                visited[new_x][new_y] = visited[cur_x][cur_y] + 1
                queue.append((new_x, new_y))
    return visited[cur_x][cur_y]  // 2


def solution(rectangle, characterX, characterY, itemX, itemY):
    graph = [[-1 for i in range(102)] for i in range(102)]
    visited = [[1 for i in range(102)] for i in range(102)]
    for rec in rectangle:
        for row in range(rec[0] * 2, rec[2] * 2 + 1):
            for col in range(rec[1] * 2, rec[3] * 2 + 1):
                if rec[0] * 2 < row < rec[2] * 2 and rec[1] * 2 < col < rec[3] * 2:
                    graph[row][col] = 0
                elif graph[row][col] != 0:
                    graph[row][col] = 1
    queue = deque([(characterX * 2, characterY * 2)])
    return bfs(graph, queue, itemX * 2, itemY * 2, visited)

File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.js
function solution(rectangle, characterX, characterY, itemX, itemY) {
    const graph = Array.from({ length: 101 }, () => Array(101).fill(0));
    
    for (const each of rectangle) {
        const [minX, minY, maxX, maxY] = each;
        
        for (let i = minX * 2; i <= maxX * 2; i++) {
            for (let j = minY * 2; j <= maxY * 2; j++) {
                if (
                    (i === minX * 2 || i === maxX * 2 || j === minY * 2 || j === maxY * 2) &&
                    graph[i][j] !== -1
                ) {
                    graph[i][j] = 1; // 경계선을 1로 표시
                } else if (i > minX * 2 && i < maxX * 2 && j > minY * 2 && j < maxY * 2) {
                    graph[i][j] = -1; // 내부는 -1로 표시
                }
            }
        }
    }
    
    const queue = [[characterX * 2, characterY * 2, 0]];
    graph[characterX * 2][characterY * 2] = 2;
    
    function BFS() {
        while (queue.length > 0) {
            const [curX, curY, distance] = queue.shift();
            
            const dx = [1, -1, 0, 0];
            const dy = [0, 0, 1, -1];
            
            for (let i = 0; i < 4; i++) {
                const newX = curX + dx[i];
                const newY = curY + dy[i];
                
                if (newX === itemX * 2 && newY === itemY * 2) {
                    return Math.floor((distance + 1) / 2);
                }
                
                if (newX >= 0 && newY >= 0 && newX < 101 && newY < 101 && graph[newX][newY] === 1) {
                    graph[newX][newY] = 2;
                    queue.push([newX, newY, distance + 1]);
                }
            }
        }
    }
    
    return BFS();
}
